README.md
# custom_save_pdf:


####  Flutter  Package  allowed you to saving pdf files in custom folder in phone storage

------------------
## # 👨‍💻 *Developed  by:*

<img alt="profile" src="https://abom.me/packages/profile.png" width="50" height="50"  style=" border-radius: 100%"/>

**Nasr Al-Rahbi [@abom_me](https://twitter.com/abom_me)**

## 👨🏻‍💻 Find me in  :
[![Twitter](https://img.shields.io/badge/Twitter-%231DA1F2.svg?logo=Twitter&logoColor=white)](https://twitter.com/abom_me)
[![Instagram](https://img.shields.io/badge/Instagram-%23E4405F.svg?logo=Instagram&logoColor=white)](https://instagram.com/abom.me)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-%230077B5.svg?logo=linkedin&logoColor=white)](https://linkedin.com/in/nasr-al-rahbi-08a573245)
[![Stack Overflow](https://img.shields.io/badge/-Stackoverflow-FE7A16?logo=stack-overflow&logoColor=white)](https://stackoverflow.com/users/19994059/nasr-al-rahbi)

---------------
<br>


**custom_save_pdf:** using the official [open_document](https://pub.dev/packages/open_document) & [path_provider](https://pub.dev/packages/path_provider) & [permission_handler](https://pub.dev/packages/permission_handler).
<br>
<br>
### ⚠️ The Requirement ⚠️ :

|             | Android   | iOS    |
| ----------- |-----------| ------ | 
| **Support** | 	SDK  29+ | iOS 11+ 




## Setup:

Pubspec changes:

```
      dependencies:
        custom_save_pdf: <last_verison>
```
Sample example:

```dart
import 'package:custom_save_pdf/custom_save_pdf.dart';

CustomPdfFolder.save(
byteList: data,
nameOfFolder: 'nasr',
openPDF: true,
onSaved: (v) {},
onError: (error, code) {
print(error);
print(code);
});
```

## Getting Started


- **You Should Follow all the instructions carefully**  
<details>
 <summary style="font-size: 15px">-----Android-----------</summary>


1. You should  make sure to set the correct `minSdkVersion` in `android/app/build.gradle` :




```groovy
android {
    defaultConfig {
        minSdkVersion 29
    }
}
```

2-You should  make sure to set the correct `compileSdkVersion` in `android/app/build.gradle` :




```groovy
android {
    compileSdkVersion 33
    ...
}
```

3-Add this  in `android/app/build.gradle` :




```xml

<!-- First add this  -->
 <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.INTERNET" />


<application
        <!-- now add this line in application   -->
android:requestLegacyExternalStorage="true"
...
        >

        <!--  add this code inside  tag application   -->
<provider
android:name="androidx.core.content.FileProvider"
android:authorities="{the pakage of app}.{the app name}.fileprovider"
android:exported="false"
android:grantUriPermissions="true">
<meta-data
        android:name="android.support.FILE_PROVIDER_PATHS"
        android:resource="@xml/provider_paths" />
</provider>

```

3- Go to android/app/src/main/res  and  create folder name it  xml then  create provider_paths.xml
```xml

<!-- Add this code in provider_paths.xml  -->
<?xml version="1.0" encoding="utf-8"?>
<paths>
    <external-path
            name="external_files"
            path="." />
</paths>
```
</details>
<br>
<br>
<br>
<details>
 <summary style="font-size: 15px">-----IOS-----------</summary>

1- add this code in info.plist:
````plist
 <key>LSSupportsOpeningDocumentsInPlace</key>
    <true/>
````
</details>


See the `example` directory for a complete sample app.


### Parameters of the CustomPdfFolder

```dart
 /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add The Uint8List From (PDF) Package
/// #### <a href="https://pub.dev/packages/pdf"> PDF Package </a>
/// #### <span style="color:green">  You Will find example in the document</span>

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### اضف كود
/// ### Uint8List
///### بإمكانك الحصول عليه من باكج
///### PDF
/// ### <a href="https://pub.dev/packages/pdf"> PDF Package </a>
/// #### <span style="color:green">  سوف تجد أمثلة في مستندات الباكج</span>

final Uint8List byteList;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
///
/// ### Write  Name for The folder, Where The PDF File Saving on it
///

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>

/// ### أكتب أسم المجلد الذي تريد حفظ الملفات بداخله
final String nameOfFolder;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ### Write  name for the file, for example:

/// #### (-- example.pdf --)
/// <span style="color:#d80000"> Without (.pdf) </span>
/// <br>
/// <span style="color:green">You can keep it empty and will make random name</span>
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### أكتب اسم للملف الذي تريد حفظه, مثال:
/// #### (-- example.pdf --)

/// <span style="color:#d80000"> بدون اضافة الامتداد (.pdf) </span>
/// <br>
/// <span style="color:green">بإمكانك تركه فارغا وسوف يقوم بإنشاء اسم عشوائي </span>

final String? fileName;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ### Here You can get the path of the pdf file as String

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا يعطيك مسار الملف كنص

final void Function(String filePath)? onSaved;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ### Here showing you any error happen with the code of the error and you can add Function if there is any errors

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا يظهر  نص و رقم الخطأ اذا حدثت اي اخطاء، و بإمكانك إعطائه امر ليفعله اذا عثر على خطأ

final void Function(String error, int errorCode)? onError;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ### keep it true to open the pdf file after save it
/// ### or false to only saving in files
/// <span style="color:green;font-size:13;font-weight: bold;">it's True by default  </span>

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// #### True
/// #### لكي ينفتح الملف بعد حفظه
/// #### False
/// #### لكي يحفظ الملف فقط بدون فتحه
bool? openPDF = true;

```


CHANGELOG.md
## 0.0.1

* uploading the package 


## 0.0.2
* fix Example project
* 
* ## 0.0.3
* Adding arabic version in readme


test/custom_save_pdf_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:custom_save_pdf/custom_save_pdf.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart
import 'package:custom_save_pdf/pdf_save/pdf_save.dart';
import 'package:flutter/material.dart';

import 'invoice_service.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  TextEditingController text = TextEditingController();

  final PdfInvoiceService service = PdfInvoiceService();

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            SizedBox(
              width: double.infinity,
              child: TextField(
                controller: text,
              ),
            ),
            SizedBox(
              height: 50,
            ),
            ElevatedButton(
              onPressed: () async {
                final data = await service.createHelloWorld(text.text);

                CustomPdfFolder.save(
                    byteList: data,
                    nameOfFolder: 'folderName',
                    openPDF: true,
                    onSaved: (v) {},
                    onError: (error, code) {
                      print(error);
                      print(code);
                    });
              },
              child: const Text("Create PDF"),
            ),
          ],
        ),
      ),
    );
  }
}


lib/custom_save_pdf.dart
library custom_save_pdf;

import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:open_document/open_document.dart';
import 'package:path_provider/path_provider.dart';
import 'package:permission_handler/permission_handler.dart';

class CustomPdfFolder {
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add The Uint8List From (PDF) Package
  /// #### <a href="https://pub.dev/packages/pdf"> PDF Package </a>
  /// #### <span style="color:green">  You Will find example in the document</span>

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### اضف كود
  /// ### Uint8List
  ///### بإمكانك الحصول عليه من باكج
  ///### PDF
  /// ### <a href="https://pub.dev/packages/pdf"> PDF Package </a>
  /// #### <span style="color:green">  سوف تجد أمثلة في مستندات الباكج</span>

  final Uint8List byteList;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  ///
  /// ### Write  Name for The folder, Where The PDF File Saving on it
  ///

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>

  /// ### أكتب أسم المجلد الذي تريد حفظ الملفات بداخله
  final String nameOfFolder;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### Write  name for the file, for example:

  /// #### (-- example.pdf --)
  /// <span style="color:#d80000"> Without (.pdf) </span>
  /// <br>
  /// <span style="color:green">You can keep it empty and will make random name</span>
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### أكتب اسم للملف الذي تريد حفظه, مثال:
  /// #### (-- example.pdf --)

  /// <span style="color:#d80000"> بدون اضافة الامتداد (.pdf) </span>
  /// <br>
  /// <span style="color:green">بإمكانك تركه فارغا وسوف يقوم بإنشاء اسم عشوائي </span>

  final String? fileName;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### Here You can get the path of the pdf file as String

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا يعطيك مسار الملف كنص

  final void Function(String filePath)? onSaved;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### Here showing you any error happen with the code of the error and you can add Function if there is any errors

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا يظهر  نص و رقم الخطأ اذا حدثت اي اخطاء، و بإمكانك إعطائه امر ليفعله اذا عثر على خطأ

  final void Function(String error, int errorCode)? onError;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### keep it true to open the pdf file after save it
  /// ### or false to only saving in files
  /// <span style="color:green;font-size:13;font-weight: bold;">it's True by default  </span>

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// #### True
  /// #### لكي ينفتح الملف بعد حفظه
  /// #### False
  /// #### لكي يحفظ الملف فقط بدون فتحه
  bool? openPDF = true;

  CustomPdfFolder({
    this.onError,
    required this.nameOfFolder,
    this.fileName,
    required this.byteList,
    this.onSaved,
    this.openPDF,
  });

  CustomPdfFolder.save({
    this.onError,
    required this.byteList,
    required this.nameOfFolder,
    this.fileName,
    this.openPDF,
    this.onSaved,
  }) {
    permission();
  }
  permission() async {
    if (await Permission.contacts.request().isGranted) {}

    await Permission.storage.request();
    await Permission.manageExternalStorage.request();

    if (await Permission.storage.isGranted) {
      savePDF();
    } else {
      onError!('User Denied Access to the storage', 101);
    }
  }

  savePDF() async {
    String newPath = '';
    Directory? directory;
    if (Platform.isAndroid) {
      directory = await getExternalStorageDirectory();
      List<String>? paths = directory?.path.split("/");
      for (int x = 1; x < paths!.length; x++) {
        String folder = paths[x];
        if (folder != "Android") {
          newPath += "/$folder";
        } else {
          break;
        }
      }
      newPath = "$newPath/$nameOfFolder";
      directory = Directory(newPath);

      await directory.create(recursive: true);
      save(directory);
    } else if (Platform.isIOS) {
      directory = await getApplicationDocumentsDirectory();

      newPath = "${directory.path}/$nameOfFolder";

      directory = Directory(newPath);
      await directory.create(recursive: true).catchError((e) {
        onError!(e.toString(), 102);
      });
      save(directory);
    }
  }

  save(
    Directory directory,
  ) async {
    var nameFile = fileName ?? '${DateTime.now()}_${getRandomString(10)}';

    var filePath = "${directory.path}/$nameFile.pdf";
    final file = File(filePath);
    await file
        .writeAsBytes(byteList)
        .then((value) => {
              onSaved!(filePath),
            })
        .catchError((onError) {
      onError!(onError.toString(), 103);
    });

    if (openPDF == true) {
      await OpenDocument.openDocument(filePath: filePath).catchError((onError) {
        onError!(onError.toString(), 104);
      });
    }
  }

  String getRandomString(int length) => String.fromCharCodes(Iterable.generate(
      length,
      (_) => 'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890'
          .codeUnitAt(Random().nextInt(
              'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890'
                  .length))));
}


example/lib/invoice_service.dart
import 'dart:typed_data';

import 'package:flutter/services.dart';
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;

class PdfInvoiceService {
  Future<Uint8List> createHelloWorld(String Text) async {
    final image =
        (await rootBundle.load("assets/logo.png")).buffer.asUint8List();

    final pdf = pw.Document();
    pdf.addPage(
      pw.Page(
        pageFormat: PdfPageFormat.a4,
        build: (pw.Context context) {
          return pw.Column(
              mainAxisAlignment: pw.MainAxisAlignment.center,
              children: [
                pw.Image(pw.MemoryImage(image),
                    width: 150, height: 150, fit: pw.BoxFit.cover),
                pw.SizedBox(height: 50),
                pw.Center(
                  child: pw.Text(Text, style: const pw.TextStyle(fontSize: 30)),
                ),
                pw.SizedBox(height: 50),
                pw.Center(
                  child: pw.Text('PDF Maker by: Nasr Al-Rahbi',
                      style: const pw.TextStyle(fontSize: 20)),
                ),
              ]);
        },
      ),
    );

    return pdf.save();
  }
}


lib/pdf_save/pdf_save.dart
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:open_document/open_document.dart';
import 'package:path_provider/path_provider.dart';
import 'package:permission_handler/permission_handler.dart';

class CustomPdfFolder {
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add The Uint8List From (PDF) Package
  /// #### <a href="https://pub.dev/packages/pdf"> PDF Package </a>
  /// #### <span style="color:green">  You Will find example in the document</span>

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### اضف كود
  /// ### Uint8List
  ///### بإمكانك الحصول عليه من باكج
  ///### PDF
  /// ### <a href="https://pub.dev/packages/pdf"> PDF Package </a>
  /// #### <span style="color:green">  سوف تجد أمثلة في مستندات الباكج</span>

  final Uint8List byteList;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  ///
  /// ### Write  Name for The folder, Where The PDF File Saving on it
  ///

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>

  /// ### أكتب أسم المجلد الذي تريد حفظ الملفات بداخله
  final String nameOfFolder;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### Write  name for the file, for example:

  /// #### (-- example.pdf --)
  /// <span style="color:#d80000"> Without (.pdf) </span>
  /// <br>
  /// <span style="color:green">You can keep it empty and will make random name</span>
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### أكتب اسم للملف الذي تريد حفظه, مثال:
  /// #### (-- example.pdf --)

  /// <span style="color:#d80000"> بدون اضافة الامتداد (.pdf) </span>
  /// <br>
  /// <span style="color:green">بإمكانك تركه فارغا وسوف يقوم بإنشاء اسم عشوائي </span>

  final String? fileName;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### Here You can get the path of the pdf file as String

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا يعطيك مسار الملف كنص

  final void Function(String filePath)? onSaved;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### Here showing you any error happen with the code of the error and you can add Function if there is any errors

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا يظهر  نص و رقم الخطأ اذا حدثت اي اخطاء، و بإمكانك إعطائه امر ليفعله اذا عثر على خطأ

  final void Function(String error, int errorCode)? onError;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ### keep it true to open the pdf file after save it
  /// ### or false to only saving in files
  /// <span style="color:green;font-size:13;font-weight: bold;">it's True by default  </span>

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// #### True
  /// #### لكي ينفتح الملف بعد حفظه
  /// #### False
  /// #### لكي يحفظ الملف فقط بدون فتحه
  bool? openPDF = true;

  CustomPdfFolder({
    this.onError,
    required this.nameOfFolder,
    this.fileName,
    required this.byteList,
    this.onSaved,
    this.openPDF,
  });

  CustomPdfFolder.save({
    this.onError,
    required this.byteList,
    required this.nameOfFolder,
    this.fileName,
    this.openPDF,
    this.onSaved,
  }) {
    permission();
  }
  permission() async {
    if (await Permission.contacts.request().isGranted) {}

    await Permission.storage.request();
    await Permission.manageExternalStorage.request();

    if (await Permission.storage.isGranted) {
      savePDF();
    } else {
      onError!('User Denied Access to the storage', 101);
    }
  }

  savePDF() async {
    String newPath = '';
    Directory? directory;
    if (Platform.isAndroid) {
      directory = await getExternalStorageDirectory();
      List<String>? paths = directory?.path.split("/");
      for (int x = 1; x < paths!.length; x++) {
        String folder = paths[x];
        if (folder != "Android") {
          newPath += "/$folder";
        } else {
          break;
        }
      }
      newPath = "$newPath/$nameOfFolder";
      directory = Directory(newPath);

      await directory.create(recursive: true);
      save(directory);
    } else if (Platform.isIOS) {
      directory = await getApplicationDocumentsDirectory();

      newPath = "${directory.path}/$nameOfFolder";

      directory = Directory(newPath);
      await directory.create(recursive: true).catchError((e) {
        onError!(e.toString(), 102);
      });
      save(directory);
    }
  }

  save(
    Directory directory,
  ) async {
    var nameFile = fileName ?? '${DateTime.now()}_${getRandomString(10)}';

    var filePath = "${directory.path}/$nameFile.pdf";
    final file = File(filePath);
    await file
        .writeAsBytes(byteList)
        .then((value) => {
              onSaved!(filePath),
            })
        .catchError((onError) {
      onError!(onError.toString(), 103);
    });

    openPDF == true
        ? await OpenDocument.openDocument(filePath: filePath)
            .catchError((onError) {
            onError!(onError.toString(), 104);
          })
        : null;
  }

  String getRandomString(int length) => String.fromCharCodes(Iterable.generate(
      length,
      (_) => 'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890'
          .codeUnitAt(Random().nextInt(
              'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890'
                  .length))));
}
